rq/job.py

```diff
@@ -1679,8 +1695,77 @@ def __init__(self, max: int, interval: Union[int, Iterable[int]] = 0):
         self.intervals = intervals
 
 
+class CallbackType(str, Enum):
+    SUCCESS = 'success'
+    FAILURE = 'failure'
+    STOPPED = 'stopped'
+
+
+@runtime_checkable
+class SuccessCallbackFunc(Protocol):
+    def __call__(self, job: Job, connection: Redis, result: Any, *args: Any, **kwargs: Any) -> Any:
+        '''
+        Callback function to be called when a job is successfully executed.
+
+        Args:
+            job (Job): the job that was executed
+            connection (Redis): the Redis connection
+            result (Any): the result of the job execution
+            *args (Any): additional positional arguments
+            **kwargs (Any): additional keyword arguments
+        '''
+
+
+@runtime_checkable
+class FailureCallbackFunc(Protocol):
+    def __call__(
+        self, job: Job, connection: Redis, type: Type[BaseException], value: BaseException, traceback: Any
+    ) -> Any:
+        '''
+        Callback function to be called when a job fails during execution.
+
+        Args:
+            job (Job): the job that was executed
+            connection (Redis): the Redis connection
+            Type (type[BaseException]): the type of the exception
+            value (BaseException): the exception instance
+            traceback (Any): the traceback of the exception
+        '''
+
+
+@runtime_checkable
+class StoppedCallbackFunc(Protocol):
+    def __call__(self, job: Job, connection: Redis) -> Any:
+        '''
+        Callback function to be called when a job is stopped before execution.
+
+        Args:
+            job (Job): the job that was executed
+            connection (Redis): the Redis connection
+        '''
+
+
+CALLBACK_PROTOCOLS = {
+    CallbackType.SUCCESS: SuccessCallbackFunc,
+    CallbackType.FAILURE: FailureCallbackFunc,
+    CallbackType.STOPPED: StoppedCallbackFunc,
+}
+
+EXPECTED_SIGNATURE = {
+    CallbackType.SUCCESS: '(job: Job, connection: Redis, result: Any, *args: Any, **kwargs: Any) -> Any',
+    CallbackType.FAILURE: (
+        '(job: Job, connection: Redis, type: Type[BaseException], ' 'value: BaseException, traceback: Any) -> Any'
+    ),
+    CallbackType.STOPPED: '(job: Job, connection: Redis) -> Any',
```

Comment:
Instead of manually typing these and hoping they keep in sync you probably should use `inspect.signature`:

```suggestion
    callback_type: str(inspect.signature(protocol.__call__))
    for callback_type, protocol in CALLBACK_PROTOCOLS.items()
```